File: waltid-applications/waltid-wallet-demo-ios/iosApp/iosApp/ViewModels/WalletInteractionClassifier.swift
import Foundation

enum WalletInteractionKind: Equatable {
    case credentialOffer
    case presentationRequest
}

enum WalletInteractionClassification: Equatable {
    case supported(kind: WalletInteractionKind, normalizedInput: String)
    case invalid(message: String)
    case unsupported(message: String)
}

func classifyWalletInteraction(_ rawInput: String) -> WalletInteractionClassification {
    let normalizedInput = rawInput.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !normalizedInput.isEmpty else {
        return .invalid(message: "Enter or scan a QR code.")
    }
    guard let separator = normalizedInput.firstIndex(of: ":"), separator != normalizedInput.startIndex else {
        return .invalid(message: "This QR code does not contain a valid wallet link.")
    }

    switch normalizedInput[..<separator].lowercased() {
    case "openid-credential-offer":
        return .supported(kind: .credentialOffer, normalizedInput: normalizedInput)
    case "openid4vp":
        return .supported(kind: .presentationRequest, normalizedInput: normalizedInput)
    default:
        return .unsupported(message: "This code is not a supported credential offer or presentation request.")
    }
}

File: waltid-applications/waltid-wallet-demo-compose/sharedLogic/src/commonMain/kotlin/id/walt/walletdemo/compose/logic/WalletInteractionClassifier.kt
package id.walt.walletdemo.compose.logic

enum class WalletInteractionKind {
    CredentialOffer,
    PresentationRequest,
}

sealed interface WalletInteractionClassification {
    data class Supported(
        val kind: WalletInteractionKind,
        val normalizedInput: String,
    ) : WalletInteractionClassification

    data class Invalid(val message: String) : WalletInteractionClassification

    data class Unsupported(val message: String) : WalletInteractionClassification
}

fun classifyWalletInteraction(rawInput: String): WalletInteractionClassification {
    val normalizedInput = rawInput.trim()
    if (normalizedInput.isEmpty()) {
        return WalletInteractionClassification.Invalid("Enter or scan a QR code.")
    }

    val scheme = normalizedInput
        .substringBefore(':', missingDelimiterValue = "")
        .lowercase()
    if (scheme.isEmpty()) {
        return WalletInteractionClassification.Invalid("This QR code does not contain a valid wallet link.")
    }

    val kind = when (scheme) {
        WalletDeepLinkScheme.CredentialOffer.scheme -> WalletInteractionKind.CredentialOffer
        WalletDeepLinkScheme.PresentationRequest.scheme -> WalletInteractionKind.PresentationRequest
        else -> null
    }
    return kind?.let { WalletInteractionClassification.Supported(it, normalizedInput) }
        ?: WalletInteractionClassification.Unsupported(
            "This code is not a supported credential offer or presentation request.",
        )
}
